### ISSUE TITLE
Missing square in RBF kernel gradient

### ISSUE THREAD

**Issue by btlorch** [Trusted-AI/adversarial-robustness-toolbox#921]
The following code snippet from the adversarial robustness toolbox computes [the gradient of an RBF kernel](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/4ca921bb336ec596a8e44f2d5b699df8fbaa9ae9/art/estimators/classification/scikitlearn.py#L1268):

```python
grad = (
    2
    * self.model._gamma
    * (-1)
    * np.exp(-self.model._gamma * np.linalg.norm(x_sample - sv, ord=2))
    * (x_sample - sv)
)
```

According to the [scikit-learn docs](https://scikit-learn.org/stable/modules/svm.html#svm-kernels), the RBF kernel is defined as `\exp(-\gamma \|x-x'\|^2)`. Using the same definition for the RBF kernel, Section 4.2.1 in [this paper by Biggio et al.](https://pralab.diee.unica.it/sites/default/files/biggio14-svm-chapter.pdf), which is also referenced in the code, describes the corresponding gradient as `-2\gamma \exp(-\gamma \| x - x' \|^2) (x - x')`.

Shouldn't the norm in the code be squared?

**Comment by beat-buesser** [comment]
Hi @btlorch Thank you very much for raising this issue! I think you are right, the derivation of the exponential function term should not change it and therefore square of `x - x'` is missing in the code.

Did you try out how much it changes the behavior of attacks?

Would you be interested to open a pull request to fix it?

**Comment by btlorch** [comment]
Thanks for the swift response, I'll be happy to open a pull request.

Here is a comparison on the binary classification toy example with projected gradient descent from `classifier_scikitlearn_SVC_LinearSVC.ipynb`.

**Projected gradient descent, current implementation, missing square in RBF kernel gradient**
![rbf_kernel_missing_square](https://user-images.githubusercontent.com/11959715/109623972-fc3cda00-7b3d-11eb-92a4-2cedd4e96a36.png)

**Projected gradient descent, after fixing the square**
![rbf_kernel_with_square](https://user-images.githubusercontent.com/11959715/109623990-fe9f3400-7b3d-11eb-8d61-bb41c1edfbdb.png)

It looks like the current implementation leads to the more desirable behavior, i.e., adversarial examples are moving towards the decision boundary. After fixing the missing square, some adversarial examples move in the direction of quickest descent. Nevertheless, this is the [behavior reported in the paper](https://pralab.diee.unica.it/sites/default/files/biggio14-svm-chapter.pdf) (see Fig. 4). In the paper, the authors propose to introduce an additional loss term that penalizes moving into low density regions. I suppose this additional loss term could be integrated into a future version of the ART.

Despite this observation, it looks like the Carlini & Wagner L2 attack is still successful in finding adversarial examples.

**C&W L2 attack, current implementation, missing square in RBF kernel gradient**
![cw_rbf_kernel_missing_square](https://user-images.githubusercontent.com/11959715/109626557-fbf20e00-7b40-11eb-8d83-e630ee5cce79.png)

**C&W L2 attack, after fixing the square**
![cw_rbf_kernel_with_square](https://user-images.githubusercontent.com/11959715/109626571-014f5880-7b41-11eb-8b66-0a82bda49077.png)

**Comment by beat-buesser** [comment]
Hi @btlorch This is very nice and interesting! Thank you very much sharing your insights and opening a pull request! I think the additional loss term against low density regions would indeed be a nice addition to future versions of ART.

**Comment by btlorch** [comment]
Thanks for merging the pull request, I am closing this issue.
